unison-cli/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -2899,24 +2865,107 @@ loadPropagateDiffDefaultPatch inputDescription maybeDest0 dest = do
+      toSplitName hq = do
+        resolvedPath <- Path.convert <$> Cli.resolveSplit' (HQ'.toName <$> hq^._1)
+        return (resolvedPath, Path.unsafeToName (Path.unsplit resolvedPath), hq^._2, hq^._3)
+  -- get the splits and names with terms and types
+  splitsNames <- traverse toSplitName typesTermsTuples
+  let toRel :: Ord ref => Set ref -> Name -> R.Relation Name ref
+      toRel setRef name = R.fromList (fmap (name,) (toList setRef))
+  let toDelete = fmap (\(_, names, types, terms) -> Names (toRel terms names) (toRel types names)) splitsNames
+  -- make sure endangered is compeletely contained in paths
+  rootNames <- Branch.toNames <$> Cli.getRootBranch0
+  -- get only once for the entire deletion set
+  let allTermsToDelete :: Set LabeledDependency
+      allTermsToDelete = Set.unions (fmap Names.labeledReferences toDelete)
+  -- get the endangered dependencies for each entity to delete
+  endangered <- Cli.runTransaction $ traverse (\targetToDelete ->
+    getEndangeredDependents targetToDelete allTermsToDelete rootNames) toDelete
+  -- If the overall dependency map is not completely empty, abort deletion
+  let endangeredDeletions = List.filter (\m -> not $ null m || Map.foldr (\s b -> null s || b ) False m ) endangered
+  if null endangeredDeletions then do
+    let deleteTypesTerms =
+          splitsNames >>= (\(split, _, types, terms) ->
+            (map (BranchUtil.makeDeleteTypeName split) . Set.toList $ types) ++
+            (map (BranchUtil.makeDeleteTermName split) . Set.toList $ terms )
+          )
+    before <- Cli.getRootBranch0
+    description <- inputDescription inputs
+    Cli.stepManyAt description deleteTypesTerms
+    case doutput of
+      DeleteOutput'Diff -> do
+        after <- Cli.getRootBranch0
+        (ppe, diff) <- diffHelper before after
+        Cli.respondNumbered (ShowDiffAfterDeleteDefinitions ppe diff)
+      DeleteOutput'NoDiff -> do
+        Cli.respond Success
+  else do
+      ppeDecl <- currentPrettyPrintEnvDecl Backend.Within
+      let combineRefs = List.foldl (Map.unionWith NESet.union) Map.empty endangeredDeletions
+      Cli.respondNumbered (CantDeleteDefinitions ppeDecl combineRefs)
+
 -- | Goal: When deleting, we might be removing the last name of a given definition (i.e. the
 -- definition is going "extinct"). In this case we may wish to take some action or warn the
 -- user about these "endangered" definitions which would now contain unnamed references.
 getEndangeredDependents ::
-  -- | Which names we want to delete
+  -- | Single target for deletion
   Names ->
+  -- | All names we want to delete (including the target)
+  Set LabeledDependency ->
   -- | All names from the root branch
   Names ->
   -- | map from references going extinct to the set of endangered dependents
   Sqlite.Transaction (Map LabeledDependency (NESet LabeledDependency))
-getEndangeredDependents namesToDelete rootNames = do
+getEndangeredDependents targetNamesToDelete allTermsToDelete rootNames = do
```

Comment:
I think this naming is confusing me; what's the difference between `targetNamesToDelete` and `allTermsToDelete`? Could we find a more descriptive name maybe?

Reply:
Updated names to `targetToDelete` meaning the single target for deletion being checked for dependencies, and `otherDesiredDeletions`, meaning the set of other entities issued in the delete command. The function's comment has been updated to make note of this argument's purpose.